README.md
# stroodlr
C++ Local-network chat client for Linux


CHANGES.md
Stroodlr (0.9):

  * Initial Development Release.
  * Create a function to convert from string to vector<char> type.
  *
  * Client:
  *
  * Use std.to_string() instead of boost.lexical_cast().
  * Remove unnecessary #includes.
  * Handle if user just hits enter with no input.
  * Get sending messages to work.
  * Get receiving messages semi-working.
  * Stop message reading from blocking the thread.
  * Make the thread exited when requested.
  * Remove old client code.
  * Do sleeps periodically to fix very high CPU usage.
  * Implement in-program help.
  * Start to separate the code into multiple files.
  * Put some files in a subdir.
  * Use pointer to vector<char> when reading data from socket (see below for explanation).
  * Put the message buses in the separate thread that handles the socket.
  * Use SendToServer() again.
  * Get exits sort of working.
  * Declutter client.cpp.
  * Add STATUS, LISTSERV commands.
  * Add support for sending messages with spaces properly.
  * Implement HISTORY command.
  * Update help info.
  * Handle cmdline args properly, add several more args.
  * Have a default server address - localhost - that doesn't always need to be specified.
  * Make HELP command output nicer.
  * Don't ask for input until connected to the server.
  * Send "CLIENTGOODBYE" instead of "Bye!" on exit.
  * Stop user from sending control messages (eg ^).
  * Stop adding spaces at the end of messages.
  * Make commands case-insensitive.
  * Add LISTMSG as an alias for LSMSG.
  * Wait for ENTER if connection fails on startup.
  * Remember what the user was typing if we lose the connection.
  * Don't freeze if user types "SEND" without a message to send.
  *
  * Server:
  *
  * Partially rewrite with boost (rather than C-sockets).
  * Use pointers to vector<char> when reading data from the socket (stops code from adding empty messages to the queue for obscure reasons).
  * Remove unneeded #includes.
  * Make a proper server outgoing message bus.
  * Send an acknowledgement, "ACK", when we receive a message from the client, and handle it in the client.
  * Print some basic info when errors occur, in both server and client.
  * Attempt to get it to reset the socket if the client exits.
  * Make servertools.cpp.
  * Remove need to always specify the portnumber (default is 50000).
  * Handle cmdline args properly.
  * Split into 2 threads.
  *
  * Both:
  *
  * Make the beginnings of a logger (designed to be similar to Python's Logging module).
  * ^ Add some setup member functions.
  * Finish the basic logging functions.
  * Try to fix an issue with mutexes.
  * Use CMake to automate builds.
  * Rewrite to avoid accessing io_service and socket classes across multiple threads (causes unpredictable crashes).
  * Get ACKs working again.
  * Test the logger.
  * Make cmdline usage functions.
  * Some initial testing with a raspberry pi.
  * Refactoring.
  * Remove obsolete function Log_Critical().
  * More refactoring.
  * Fix some segfaults.
  * Add logging messages.
  * Add level support to the logger.
  * Use mutexes in the logger to prevent corruption when there's multithreaded logging happening.
  * Add missing #include <stdexcept>.
  * Test with raspberry pi again.
  * Move cmdline args handlers into separate files.
  * Use const and & where possible to protect and avoid copying data.
  * Start work on a new Sockets class.
  * Split it into 2 classes.
  * Get ClientSocket working.
  * Fix segfault on exit.
  * Get ServerSocket working.
  * Remove old socket functions.
  * Fix a problem with SevrerSockets.
  * Merge the socket classes.
  * Fix weird chars in the logfile.
  * ^ Fix using string comparisons with socket data.
  * Handle CTRL-C.
  * Use special ASCII code \x06 for ACK instead of "ACK".
  * Handle connection errors properly.
  * Handle CTRL-C better.
  * Add the send/receive functions into the Sockets class.
  * Begin to make the Sockets class more independent and generic.
  * ^ Fix a related stack corruption issue.
  * Make Sockets self-managed.
  * Make Sockets.Reset().
  * Refactoring.
  * Make Sockets class handle reconnection itself.
  * Handle losing the connection better.
  * Add more logging messages.
  * Sockets: Allow disabling console output.
  * Clean up CMakeLists.txt.
  * Make a static-library out of the common files to avoid compiling them twice each time.
  * Clean up source code and remove unneeded #includes.
  * Remove a deprecated function.
  * Move SendToServer() to Sockets.SendToPeer().
  * More cleanup.
  * Do "#pragma once" in headers to prevent re-importing.
  * CMake build system: Enable turning debugging symbols and optimisations on and off with cmdline options.
  * Test that reconnecting works.
  * Protect some of the Sockets functions by making them private.
  * Add more logging messages in Sockets functions.
  * Re-do the signal handlers.
  * Add build+dist dirs to the github repo.
  * Remove unwanted conio.h stuff.
  * Fix CMakeLists.txt glitchy compilation options.


src/client.cpp
/*
Stroodlr Client Version 0.9
This file is part of Stroodlr.
Copyright (C) 2017 Hamish McIntyre-Bhatty
Stroodlr is free software: you can redistribute it and/or modify it
under the terms of the GNU General Public License version 3 or,
at your option, any later version.

Stroodlr is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with Stroodlr.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <iostream>
#include <deque>
#include <vector>
#include <string>
#include <chrono>
#include <thread>
#include <cctype> //Character handling functions.
#include <signal.h> //POSIX-only. *** Try to find an alternative solution - might not be thread-safe *** 
#include <stdexcept>

//Custom headers.
#include "../include/tools.h"
#include "../include/loggertools.h"
#include "../include/clienttools.h"
#include "../include/sockettools.h"

//Define keys.
#define UP_KEY 72
#define DOWN_KEY 80
#define LEFT_KEY 75
#define RIGHT_KEY 77

using std::string;
using std::vector;
using std::deque;

//Logger.
Logging Logger;

bool ReadyForTransmission = false;

void Usage() {
    //Prints cmdline options.
    std::cout << "Usage: stroodlrc [OPTION]" << std::endl << std::endl << std::endl;
    std::cout << "Options:" << std::endl << std::endl;
    std::cout << "        -h, --help:               Show this help message." << std::endl;
    std::cout << "        -a, --serveraddress:      Specify the server address (if unspecified, assumed to be localhost)." << std::endl;
    std::cout << "        -q, --quiet:              Show only warnings, errors and critical errors in the log file." << std::endl;
    std::cout << "                                  Very unhelpful for debugging, and not recommended." << std::endl;
    std::cout << "        -v, --verbose:            Enable logging of info messages, as well as warnings, errors and critical errors." << std::endl;
    std::cout << "                                  Best choice if there is little disk space available. The default." << std::endl;
    std::cout << "        -d, --debug:              Log lots of boring debug messages. Usually used for diagnostic purposes." << std::endl << std::endl;
    std::cout << "Stroodlr "+Version+" is released under the GNU GPL Version 3" << std::endl;
    std::cout << "Copyright (C) Hamish McIntyre-Bhatty 2017" << std::endl;
    exit(0);

}

int main(int argc, char* argv[])
{
    //Setup the logger. *** Handle exceptions ***
    Logger.SetName("Stroodlr Client "+Version);
    Logger.SetDateTimeFormat("%d/%m/%Y %I:%M:%S %p");
    Logger.SetFileName("/tmp/stroodlrc.log");
    Logger.SetStyle("Time Name Level");

    std::cout << "Stroodlr Client " << Version << " Starting..." << std::endl;
    Logger.Info("Stroodlr Client "+Version+" Starting...");

    //Setup.
    Logger.SetLevel("Info");
    string ServerAddress = "localhost";
    int PortNumber = 50000;

    //Vars to hold temporary data *** Clean up ***
    string command;
    vector<string> splitcommand;
    deque<string> UserInput;
    string abouttosend;
    string UpperCommand;
    string KeepWriting = "";
    string OldCommand = "";

    //Parse the commandline options.
    try {
        ServerAddress = ParseCmdlineOptions(ServerAddress, argc, argv);

    } catch (std::runtime_error const& e) {
        //Print the error, print usage and exit.
        std::cerr << e.what() << std::endl;
        Usage();

    }

    //Setup socket.
    Sockets Plug("Plug");

    Plug.SetPortNumber(PortNumber);
    Plug.SetServerAddress(ServerAddress);

    Plug.StartHandler();

    Logger.Info("main(): Waiting for connection to server...");
    std::cout << std::endl << "Connecting to server..." << std::endl;

    //Wait until we're connected, or requested to exit.
    while (!Plug.IsReady() && !Plug.HandlerHasExited()) std::this_thread::sleep_for(std::chrono::milliseconds(100));

    if (!Plug.IsReady()) {
        //Couldn't connect to server.
        Logger.Critical("Couldn't connect to server! Exiting...");

        //Wait for ENTER key before exiting.
        getline(std::cin, command);

        exit(1);

    }

    Logger.Info("main(): Connected...");
    std::cout << "Connected!" << std::endl;

    //Setup signal handler.
    signal(SIGINT, RequestExit);

    //Greet user and start waiting for commands.
    //Display greeting.
    std::cout << std::endl << "Welcome to Stroodlr, the local network chat client!" << std::endl;
    std::cout << "For help, type \"HELP\"" << std::endl;
    std::cout << "To quit, type \"QUIT\", \"Q\", \"EXIT\", or press CTRL-D" << std::endl << std::endl;

    //Main input loop.
    while (!::RequestedExit) {
        //Check that we're still connected.
        if (!Plug.IsReady()) {
            Logger.Info("main(): Server has disconnected. Waiting for the socket to reconnect...");

            //Deregister signal handler, so we can exit if we get stuck while connecting.
            signal(SIGINT, SIG_DFL);

            //Wait until we're connected or have to exit because of a connection error..
            while (!Plug.IsReady() && !Plug.HandlerHasExited()) std::this_thread::sleep_for(std::chrono::milliseconds(100));

            if (Plug.HandlerHasExited()) {
                //Couldn't reconnect to client.
                Logger.Critical("Couldn't reconnect to server! Exiting...");

                exit(1);

            }

            //Setup signal handler, as we are now reconnected.
            signal(SIGINT, RequestExit);

        }

        Logger.Debug("main(): Checking for new messages...");
        CheckForMessages(&Plug);

        //If the user was midway through typing something and got interrupted, let them know what it was, and push it back onto the input stream.
        if (command != "") {
            std::cout << "You were just typing: " << command << std::endl;

            while (KeepWriting == "") {
                std::cout << "Would you like to keep writing this command (y/n)? ";
                getline(std::cin, KeepWriting);

                //Convert KeepWriting to upper case.
                for (int i = 0; i < KeepWriting.size(); i++) {
                    KeepWriting[i] = toupper(KeepWriting[i]);

                }

                if (KeepWriting == "Y" || KeepWriting == "YES") {
                    OldCommand = command;

                } else if (KeepWriting == "N" || KeepWriting == "NO") {
                    OldCommand = "";

                } else {
                    KeepWriting = "";

                }
            }

            command = "";
            KeepWriting = "";

        }

        //Input prompt.
        std::cout << ">>>";

        //If there's an old command we need to write, do it now.
        if (OldCommand != "") {
            std::cout << OldCommand;

        }

        Logger.Debug("main(): Waiting for user input...");
        getline(std::cin, command);

        //If there's an old command, combine it with whatever we typed here.
        if (OldCommand != "") {
            command = OldCommand+command;

        }

        //It's possible that at this point we have lost the connection and failed to reconnect/reconnected, so check.
        if (Plug.HandlerHasExited()) {
            //Connection lost and couldn't reconnect to client.
            Logger.Critical("Couldn't reconnect to server! Exiting...");

            exit(1);

        } else if (Plug.JustReconnected()) {
            //We just reconnected. Go back to the start of the loop.
            continue;

        } else if (!Plug.IsReady()) {
            //Trying to reconnect.
            continue;

        }

        splitcommand = split(command, " ");

        //Convert the first part of the text (the "command") to upper case.
        for (int i = 0; i < splitcommand[0].size(); i++) {
            splitcommand[0][i] = static_cast<char>(toupper(splitcommand[0][i]));

        }

        //Handle invalid/quit/no input from user.
        if (!std::cin) {
            //Invalid input.
            Logger.Warning("main(): Invalid input. Asking for more input...");
            std::cout << std::endl << "Invalid input!" << std::endl;
            std::cin.clear();
            command = "";

            continue;

        } else if ((splitcommand[0] == "QUIT") || (splitcommand[0] == "Q") || (splitcommand[0] == "EXIT")) {
            //User has requested that we exit.
            Logger.Info("main(): User requested an exit...");
            break;

        } else if (splitcommand[0] == "") {
            //No input, just hit enter key. Print ">>>" again.
            Logger.Debug("main(): No Input. Asking for more input...");
            continue;

        } else {
            //Input is valid. Save it so we can recall later.
            Logger.Debug("main(): Valid input. Saving to UserInput so we can recall it later if needed...");
            UserInput.push_back(command);

        }

        //Handle "Proper" input.
        if (splitcommand[0] == "HISTORY") {
            Logger.Info("main(): Showing history...");
            ShowHistory(UserInput);

        } else if (splitcommand[0] == "STATUS") {
            Logger.Info("main(): Showing status...");
            ShowStatus();

        } else if (splitcommand[0] == "LISTSERV") {
            Logger.Info("main(): Listing connected servers...");
            ListConnectedServers();

        } else if (splitcommand[0] == "LSMSG" || splitcommand[0] == "LISTMSG") {
            Logger.Info("main(): Listing messages...");
            ListMessages(&Plug);

        } else if (splitcommand[0] == "HELP") {
            Logger.Info("main(): Showing help...");
            ShowHelp();

        } else if (splitcommand[0] == "SEND") {
            //Check if there was actually a message specified.
            if (command == "SEND") {
                //Refuse to send it, because there's no message.
                std::cout << std::endl << "You didn't specify a message to send! Usage: SEND <msg>." << std::endl << std::endl;
                command = "";

                continue;

            }

            //Send a message.
            Logger.Debug("main(): Preparing to send a message...");

            //Get the 2nd element and onwards, assemble into a string.
            Logger.Debug("main(): Assembling relevant parts of input into a string containing the message...");
            splitcommand.erase(splitcommand.begin(), splitcommand.begin() + 1);

            //Assemble into a string.
            abouttosend = "";

            for (int i = 0; i < splitcommand.size(); i++) {
                if (i != splitcommand.size() - 1) {
                    abouttosend = abouttosend + splitcommand[i] + " ";

                } else {
                    abouttosend = abouttosend + splitcommand[i];

                }
            }

            //Check this isn't a control message used for client-server communication.
            if (abouttosend == "PEERGOODBYE") {
                //Refuse to send it.
                Logger.Error("main(): Won't send control message PEERGOODBYE. Warning user...");
                std::cout << std::endl << "You cannot send control messages." << std::endl << std::endl;
                command = "";

                continue;

            }

            //Send it.
            Logger.Info("main(): Sending the message...");
            Plug.SendToPeer(ConvertToVectorChar(abouttosend));
            Logger.Info("main(): Done.");

        } else {
            Logger.Error("main(): Invalid command.");
            std::cout << "ERROR: Command not recognised. Type \"HELP\" for commands." << std::endl;
        }

        //Trim history below 100 items if needed (stops it from consuming too much RAM).
        if (UserInput.size() > 99) {
            UserInput.pop_front();

        }

        //Reset command to "".
        command = "";
        OldCommand = "";

    }

    //Exit if we broke out of the loop.
    Logger.Info("main(): Done. Saying goodbye to user and requesting that all threads exit...");

    std::cout << std::endl << "Bye!" << std::endl;
    ::RequestedExit = true;

    Plug.RequestHandlerExit();
    Plug.WaitForHandlerToExit();

    Logger.Info("main(): All threads have exited. Exiting...");
    std::cout << "Exiting..." << std::endl;

    return 0;
}
